# SMS spam filter

In this project we'll build a spam filter for SMS messages, capable of distinguishing messages between spam and non spam. 

We'll do this using as reference a set of messages previously classified as spam or non spam by humans.

## Choosing an algorithm for a classification problem

Being this a classical binary classification problem, the **Naive Bayes algorithm** should do the job well, so further on this work, we'll use it.

Let's load the libraries we need to do the job.

In [1]:
import pandas as pd
import numpy as np
import re

## The SMS Spam Collection

Now, let's load the spam collection file, convert it to a Pandas Dataframe and check it briefly.

In [2]:
colec = pd.read_csv('SMSSpamCollection', sep='\t', header=None)

In [3]:
colec

0                                                  1
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
5     spam  FreeMsg Hey there darling it's been 3 week's n...
6      ham  Even my brother is not like to speak with me. ...
7      ham  As per your request 'Melle Melle (Oru Minnamin...
8     spam  WINNER!! As a valued network customer you have...
9     spam  Had your mobile 11 months or more? U R entitle...
10     ham  I'm gonna be home soon and i don't want to tal...
11    spam  SIX chances to win CASH! From 100 to 20,000 po...
12    spam  URGENT! You have won a 1 week FREE membership ...
13     ham  I've been searching for the right words to tha...
14     ham                I HAVE A DATE ON SUNDAY WITH WILL!!
15    spam  XXXMobileMovieClub: To use your credit, click ...
16     ham                         Oh k...i'm watching here:)
17     ham  Eh u remember how 2 spell his name... Yes i di...
18     ham  Fine if that’s the way u feel. That’s the way ...
19    spam  England v Macedonia - dont miss the goals/team...
20     ham          Is that seriously how you spell his name?
21     ham    I‘m going to try for 2 months ha ha only joking
22     ham  So ü pay first lar... Then when is da stock co...
23     ham  Aft i finish my lunch then i go str down lor. ...
24     ham  Ffffffffff. Alright no way I can meet up with ...
25     ham  Just forced myself to eat a slice. I'm really ...
26     ham                     Lol your always so convincing.
27     ham  Did you catch the bus ? Are you frying an egg ...
28     ham  I'm back &amp; we're packing the car now, I'll...
29     ham  Ahhh. Work. I vaguely remember that! What does...
...    ...                                                ...
5542   ham           Armand says get your ass over to epsilon
5543   ham             U still havent got urself a jacket ah?
5544   ham  I'm taking derek &amp; taylor to walmart, if I...
5545   ham      Hi its in durban are you still on this number
5546   ham         Ic. There are a lotta childporn cars then.
5547  spam  Had your contract mobile 11 Mnths? Latest Moto...
5548   ham                 No, I was trying it all weekend ;V
5549   ham  You know, wot people wear. T shirts, jumpers, ...
5550   ham        Cool, what time you think you can get here?
5551   ham  Wen did you get so spiritual and deep. That's ...
5552   ham  Have a safe trip to Nigeria. Wish you happines...
5553   ham                        Hahaha..use your brain dear
5554   ham  Well keep in mind I've only got enough gas for...
5555   ham  Yeh. Indians was nice. Tho it did kane me off ...
5556   ham  Yes i have. So that's why u texted. Pshew...mi...
5557   ham  No. I meant the calculation is the same. That ...
5558   ham                             Sorry, I'll call later
5559   ham  if you aren't here in the next  &lt;#&gt;  hou...
5560   ham                  Anything lor. Juz both of us lor.
5561   ham  Get me out of this dump heap. My mom decided t...
5562   ham  Ok lor... Sony ericsson salesman... I ask shuh...
5563   ham                                Ard 6 like dat lor.
5564   ham  Why don't you wait 'til at least wednesday to ...
5565   ham                                       Huh y lei...
5566  spam  REMINDER FROM O2: To get 2.50 pounds free call...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

These columns could have more intuitive names, since they are named "0" and "1". Let's change them to "Label" and "SMS'.

In [4]:
colec.columns = ['Label', 'SMS']

Checking below if everything is ok with the Dataframe and its length.

In [5]:
colec.head()

Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [6]:
colec.shape

(5572, 2)

We have 5.572 rows and 2 columns on this collection. Let's check how the entries are distributed between 'ham' and 'spam', which means "non spam" and "spam" and see these percentuals.

In [7]:
ham_perc = ((len(colec[colec['Label'] == 'ham']))/len(colec))*100
spam_perc = ((len(colec[colec['Label'] == 'spam']))/len(colec))*100

In [8]:
ham_perc

86.59368269921033

In [9]:
spam_perc

13.406317300789663

This dataset is 86,59% non spam and 13,5% spam SMS.

## Creating a training set and a test set

Using this collection, we'll now create two separeted sets, one for training our filter, and other to test it. Since all these messages are already pre classified as spam or no spam, this is useful for us to check the accuracy of our solution.

The training set will have 80% of the messages, and the test set will have 20% of the total.

In [10]:
# Randomize the dataset
data_randomized = colec.sample(frac=1, random_state=1)

# Calculate index for split
training_test_index = round(len(data_randomized) * 0.8)

# Training/Test split
training_set = data_randomized[:training_test_index].reset_index(drop=True)
test_set = data_randomized[training_test_index:].reset_index(drop=True)

print(training_set.shape)
print(test_set.shape)

(4458, 2)
(1114, 2)


We'll just verify if the division was made correctly.

In [11]:
len(training_set)

4458

In [12]:
len(test_set)

1114

Everything looks right. We'll now check if the ratio spam:ham from the original dataset was kept. If nothing gone wrong, we should see the spam percentage below.

In [13]:
(len(training_set[training_set['Label'] == 'spam'])/len(training_set))*100

13.458950201884253

The percentages match. We'll now do the data cleansing on the training_set.

## Data cleansing on the training set

To standardize this this training set, we'll need to do some cleaning to remove punctuactions and to make it all lower case.

In [14]:
training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')
training_set['SMS'] = training_set['SMS'].str.lower()

In [15]:
training_set.head()

Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

This dataset is clean and ready for further work.

## Creating a vocabulary

To make a vocabulary and check the probability for every word on our dataset, we'll need to register and count every word we have. To do so, we'll loop through every row of the dataset, and through every word on every row.

In [16]:
training_set['SMS'] = training_set['SMS'].str.split()

vocabulary = []

for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)

In [17]:
training_set['SMS'].head()

0                    [yep, by, the, pretty, sculpture]
1    [yes, princess, are, you, going, to, make, me,...
2                      [welp, apparently, he, retired]
3                                             [havent]
4    [i, forgot, 2, ask, ü, all, smth, there, s, a,...
Name: SMS, dtype: object

Now, we'll use a trick to remove duplicates from our vocabulary. We'll transform the list into a set, and into a list again.

In [18]:
vocabulary = list(set(vocabulary))

In [19]:
len(vocabulary)

7783

Our vocabulary has 7.783 unique words.

## Creating a table for counting words frequency

Now we have a vocabulary, we need to check the frequency for every word. This will be done by creating a table where each row is represents an SMS, and each column represents a word. Each cell will have a number, representing how many occurences of that word the row has.

In [20]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

Let's now transform this dictionary into a dataframe.

In [21]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

0  00  000  000pes  008704050406  0089  01223585334  02  0207  02072069400  \
0  0   0    0       0             0     0            0   0     0            0   
1  0   0    0       0             0     0            0   0     0            0   
2  0   0    0       0             0     0            0   0     0            0   
3  0   0    0       0             0     0            0   0     0            0   
4  0   0    0       0             0     0            0   0     0            0   

  ...  zindgi  zoe  zogtorius  zouk  zyada  é  ú1  ü  〨ud  鈥  
0 ...       0    0          0     0      0  0   0  0    0  0  
1 ...       0    0          0     0      0  0   0  0    0  0  
2 ...       0    0          0     0      0  0   0  0    0  0  
3 ...       0    0          0     0      0  0   0  0    0  0  
4 ...       0    0          0     0      0  0   0  2    0  0  

[5 rows x 7783 columns]

Our dataframe has 4458 rows (one for each sms) and 7783 columns. We'll now concatenate this word count with the training dataset to get the index and the complete SMS of each row.

In [22]:
training_set_clean = pd.concat([training_set, word_counts], axis=1)

In [23]:
training_set_clean.shape

(4458, 7785)

In [24]:
training_set_clean = training_set_clean[training_set_clean['Label'].notnull()]

In [25]:
training_set_clean.shape

(4458, 7785)

Our new dataframe has 5341 rows (sms) and 7785 columns (words), precisely 2 more than the word counts dataframe.

## Let's now calculate our constant values for calculation probabilities, which are:
* p_spam (probability of finding an spam sms)
* p_ham (probability of finding a non-spam sms)
* n_spam (number of words in all spam sms)
* n_ham (number of words in all non spam sms)
* n_vocabulary (number of total words)
* alpha (the Laplace Smoothing factor)

In [26]:
spam_messages = training_set_clean[training_set_clean['Label'] == 'spam']
ham_messages = training_set_clean[training_set_clean['Label'] == 'ham']

In [27]:
p_spam = len(spam_messages)/len(training_set_clean)
p_spam

0.13458950201884254

In [28]:
len(training_set_clean)

4458

In [29]:
p_ham = len(ham_messages)/len(training_set_clean)
p_ham

0.8654104979811574

In [30]:
n_spam = spam_messages['SMS'].apply(len).sum()
n_spam

15190

In [31]:
n_ham = ham_messages['SMS'].apply(len).sum()
n_ham

57237

In [32]:
n_vocabulary = len(vocabulary)

n_vocabulary

7783

In [33]:
alpha = 1

In [34]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

### Extracting word probabilities, creating parameters

Now we have our constants, it's time to calculate the probability for every word on our dictionary, spam and non spam. We'll do this by using a dictionary, where each key is a word and each number is a value.

We'll start with the probabilities for words being spam.

In [35]:
parameters_spam = {unique_word:0 for unique_word in vocabulary}

In [36]:
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()
    p_word_given_spam = (n_word_given_spam + alpha)/(n_spam + alpha * n_vocabulary)
    parameters_spam[word] = p_word_given_spam

And now, probability for words not being spam.

In [37]:
parameters_ham = {unique_word:0 for unique_word in vocabulary}

In [38]:
for word in vocabulary:
    n_word_given_ham = ham_messages[word].sum()
    p_word_given_ham = (n_word_given_ham + alpha)/(n_ham + alpha * n_vocabulary)
    parameters_ham[word] = p_word_given_ham

### Creating the filter itself

In [51]:
def classify(message):
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
   
    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    if p_spam_given_message > p_ham_given_message:
        return 'spam'
    if p_spam_given_message == p_ham_given_message:
        return 'needs human classification'

### Testing the filter

We'll now test the filter with two sample messages:
* WINNER!! This is the secret code to unlock the money: C3421.
* Sounds good, Tom, then see u there

Let's see if the results look good enough. The filter will return "spam" for a spam message, and "ham" for a non-spam message.

In [40]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

'spam'

In [52]:
classify('Sounds good, Tom, then see u there')

'ham'

Looks good. Let's check the accuracy of the filter, using the test set, which has pre-classified values to compare.

In [42]:
test_set['predicted'] = test_set['SMS'].apply(classify)

In [43]:
test_set.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

In [55]:
correct = 0
total = len(test_set)

Below, we'll check the number of rows which the pre-classification is equal to the generated by the filter

In [57]:
corr = len(test_set[test_set['Label'] == test_set['predicted']])
corr

1100

And now, the rows with wrong classifications according to the original set.

In [58]:
wrong = len(test_set[test_set['Label'] != test_set['predicted']])
wrong

14

Below, we have the percentage of correct answers given by our filter.

In [59]:
(corr/total)*100

98.74326750448833

And now, lets check all the wrong predictions.

In [60]:
test_set[test_set['Label'] != test_set['predicted']]

Label                                                SMS  \
114  spam  Not heard from U4 a while. Call me now am here...   
135  spam  More people are dogging in your area now. Call...   
152   ham                  Unlimited texts. Limited minutes.   
159   ham                                       26th OF JULY   
284   ham                             Nokia phone is lovly..   
293   ham  A Boy loved a gal. He propsd bt she didnt mind...   
302   ham                   No calls..messages..missed calls   
319   ham  We have sent JD for Customer Service cum Accou...   
504  spam  Oh my god! I've found your number again! I'm s...   
546  spam  Hi babe its Chloe, how r u? I was smashed on s...   
741  spam  0A$NETWORKS allow companies to bill for SMS, s...   
876  spam           RCT' THNQ Adrian for U text. Rgds Vatian   
885  spam                                      2/2 146tf150p   
953  spam  Hello. We need some posh birds and chaps to us...   

                      predicted  
114                         ham  
135                         ham  
152                        spam  
159                        spam  
284                        spam  
293  needs human classification  
302                        spam  
319                        spam  
504                         ham  
546                         ham  
741                         ham  
876                         ham  
885                         ham  
953                         ham

With this results and performance, our spam filter is fully functional and good to go.